tezos/new_context/src/gc/worker.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    collections::{BTreeMap, VecDeque},
+    sync::Arc,
+};
+
+use crossbeam_channel::Receiver;
+
+use crate::{kv_store::HashId, working_tree::Entry};
+
+use tezos_spsc::Producer;
+
+pub(crate) const PRESERVE_CYCLE_COUNT: usize = 7;
+
+pub(crate) struct GCThread {
+    pub(crate) cycles: Cycles,
+    pub(crate) free_ids: Producer<HashId>,
+    pub(crate) recv: Receiver<Command>,
+    pub(crate) pending: Vec<HashId>,
+}
+
+pub(crate) enum Command {
+    StartNewCycle {
+        values_in_cycle: BTreeMap<HashId, Option<Arc<[u8]>>>,
+        new_ids: Vec<HashId>,
+    },
+    MarkReused {
+        reused: Vec<HashId>,
+    },
+}
+
+pub(crate) struct Cycles {
+    list: VecDeque<BTreeMap<HashId, Option<Arc<[u8]>>>>,
+}
+
+impl Default for Cycles {
+    fn default() -> Self {
+        let mut list = VecDeque::with_capacity(PRESERVE_CYCLE_COUNT);
+
+        for _ in 0..PRESERVE_CYCLE_COUNT {
+            list.push_back(Default::default());
+        }
+
+        Self { list }
+    }
+}
+
+impl Cycles {
+    fn move_to_last_cycle(&mut self, hash_id: HashId) -> Option<Arc<[u8]>> {
+        let len = self.list.len();
+        let mut value = None;
+
+        for store in &mut self.list.iter_mut().take(len - 1) {
+            if let Some(item) = store.remove(&hash_id).flatten() {
+                value = Some(item);
+            };
+        }
+
+        let value = value?;
+        if let Some(v) = self.list.back_mut() {
+            v.insert(hash_id, Some(Arc::clone(&value)));
+        } else {
+            eprintln!("GC: Failed to insert value in Cycles")
+        }
+
+        return Some(value);
+    }
+
+    fn roll(&mut self, new_cycle: BTreeMap<HashId, Option<Arc<[u8]>>>) -> Vec<HashId> {
+        let unused = self.list.pop_front().unwrap_or_else(BTreeMap::new);
+        self.list.push_back(new_cycle);
+
+        let mut vec = Vec::with_capacity(unused.len());
+        for id in unused.keys() {
+            vec.push(*id);
+        }
+        vec
+    }
+}
+
+impl GCThread {
+    pub(crate) fn run(mut self) {
+        while let Ok(msg) = self.recv.recv() {
```

Comment:
@sebastiencs what would be an error result here? thinking that `loop` + `match` with logging + break on `Err`, or maybe even have `run` return `Result<(), AnError>` would be better. This is one of those things that if it fails silently, would be super hard to figure out.